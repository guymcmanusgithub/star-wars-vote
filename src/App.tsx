import React, { useEffect, useState } from 'react';
import axios, { AxiosResponse } from 'axios'
import { MovieTable } from './components/movieTable/MovieTable';
import { Container, H1, P, Span } from './App.style'

export const swapiApi:string = 'https://swapi.dev/api/films/';
 
export const fetchData:ApiCall = async () => {
  const url = `${swapiApi}`;
  return await axios.get(url);
};

function App() {
  const [movieInfo, setMovieInfo] = useState<Array<Movie>>([])
  const [totalVotes, setTotalVotes] = useState<number>(0)
  const [loading, setLoading] = useState<boolean>(true)
  const [apiError, setApiError] = useState<boolean>(false)
  const [errorMessage, setErrorMessage] = useState<string>('')

  useEffect(() => {
    return fetchData()
    .then((data:AxiosResponse<SWAPIData>) => { 
      const {data:{results}} = data
      
      const newMovieArray:Array<Movie> = results.map((movie:Results) => (
        {
          title: movie.title,
          votes: 0,
          date: movie.release_date
        }
      ))
      // The data from SWAPI already comes back sorted by date, so the below function is simply a failsafe if the api decides to return the data not sorted by date in the future.
      newMovieArray.sort((a:Movie, b:Movie) => {
        return (a.date < b.date) ? -1 : ((a.date > b.date) ? 1 : 0);
      });
      setMovieInfo(newMovieArray)
      setLoading(false)
      setApiError(false)
    })
    .catch((err: any) => { 
      setLoading(false)
      setApiError(true)
      setErrorMessage(err.message)
    });
  },[])

  const votesReducer:number = movieInfo.reduce((acc:number, movie:Movie) => acc + movie.votes, 1)

  const voteClick:VoteClick = (title) => {
    const newMovies:React.SetStateAction<Movie[]> = movieInfo.map(movie => {
      let {votes} = movie
      if(movie.title === title) {
        return {...movie, votes: (votes + 1)}
      } 
      return movie 
    })
    setMovieInfo(newMovies);
    setTotalVotes(votesReducer);
  }

  if (loading){
    return <P data-testid="loading">I find your lack of faith disturbing.  The app is <strong>Loading...</strong></P>
  }

  if (apiError && !loading) {
    return <P data-testid="api-error">You would be wise to surrender. An error has occurred: {errorMessage}</P>
  }

  return (
    <Container data-testid="app-container">
      {!loading && !apiError &&
        <> 
          <H1>Vote Wars</H1>
          <MovieTable movieInfo={movieInfo} voteClick={voteClick}/>
          <P>Total Votes: <Span>{totalVotes}</Span></P>
        </>
      }
    </Container>
  );
}

export default App;