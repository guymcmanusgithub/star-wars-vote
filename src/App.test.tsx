import React from 'react';
import App from './App';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, cleanup, findByText, fireEvent,} from '@testing-library/react';
import { sWAPIdata } from './mockTestData';
import axios from 'axios';

jest.mock('axios')

beforeEach(() => {
  jest.clearAllMocks();
})

afterEach(cleanup)
  
describe('<App/> component shows three possible components', () => {

  it('shows a loading screen before returning API data', () => {
      render(<App/>)
        const element = screen.getByTestId('loading')
        expect(element).toMatchSnapshot()
        expect(element).toBeVisible()
    })

  it('shows an error screen when erroring', async() => {
    jest.spyOn(axios, 'get').mockRejectedValue({
      error: "There has been an error"
    })
      
    const { findByTestId} = render(<App/>)
    const element =  await findByTestId('api-error')
    expect(element).toMatchSnapshot()
    const errorMessage =  await findByText(element, 'You would be wise to surrender. An error has occurred:')
    expect(errorMessage).toBeTruthy()
  })

  it('calls the get method from axios', async () => {
    jest.spyOn(axios, 'get').mockResolvedValue(
      sWAPIdata
    )
      
    const {findByTestId} = render(<App/>)
    const container =  await findByTestId('app-container')
    expect(container).toMatchSnapshot()
    const voteWars =  await findByText(container, 'Vote Wars')
    expect(voteWars).toBeTruthy()
  })
       
})

describe('<App/> functions', () => {
  it('adds a vote when the vote button is clicked', async () => {
    jest.spyOn(axios, 'get').mockResolvedValue(
      sWAPIdata
    )
      
    const {findByTestId} = render(<App/>)
    await findByTestId('app-container')

    const initialVotes = screen.getByTestId('votes-Return of the Jedi')
    expect(initialVotes).toHaveTextContent('0')

    const button = screen.getByTestId('button-Return of the Jedi')
    expect(fireEvent.click(button)).toBeTruthy()

    const updatedVotes = screen.getByTestId('votes-Return of the Jedi')
    expect(updatedVotes).toHaveTextContent('1')
    expect(updatedVotes).not.toHaveTextContent('3')
  })
})

