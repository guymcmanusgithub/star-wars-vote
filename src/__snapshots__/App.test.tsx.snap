// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> component shows three possible components calls the get method from axios 1`] = `
<div
  class="sc-fubCfw ewlTFH"
  data-testid="app-container"
>
  <h1
    class="sc-pFZIQ dVfauT"
  >
    Vote Wars
  </h1>
  <table
    class="sc-bdfBwQ kSOtiz"
  >
    <caption
      class="sc-gsTCUz iNWYbO"
    >
      Vote for your favorite Star Wars movie:
    </caption>
    <thead
      class="sc-dlfnbm"
    >
      <tr
        class="sc-eCssSg gSspkk"
      >
        <th
          class="sc-jSgupP kXXhUQ"
          id="movie-title"
        >
          Movie Title
        </th>
        <th
          class="sc-jSgupP kXXhUQ"
          id="vote"
        >
          Vote
        </th>
        <th
          class="sc-jSgupP kXXhUQ"
          id="total-votes"
        >
          Total Votes
        </th>
      </tr>
    </thead>
    <tbody
      class="sc-hKgILt"
    >
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          A New Hope
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-A New Hope"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-A New Hope"
          headers="total-votes"
        >
          0
        </td>
      </tr>
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          The Empire Strikes Back
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-The Empire Strikes Back"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-The Empire Strikes Back"
          headers="total-votes"
        >
          0
        </td>
      </tr>
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          Return of the Jedi
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-Return of the Jedi"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-Return of the Jedi"
          headers="total-votes"
        >
          0
        </td>
      </tr>
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          The Phantom Menace
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-The Phantom Menace"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-The Phantom Menace"
          headers="total-votes"
        >
          0
        </td>
      </tr>
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          Attack of the Clones
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-Attack of the Clones"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-Attack of the Clones"
          headers="total-votes"
        >
          0
        </td>
      </tr>
      <tr
        class="sc-eCssSg gSspkk"
      >
        <td
          class="sc-gKsewC gbAxPW"
          headers="movie-title"
        >
          Revenge of the Sith
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          headers="vote"
        >
          <button
            aria-label="vote-button"
            class="sc-iBPRYJ cdPLuh"
            data-testid="button-Revenge of the Sith"
          >
            vote
          </button>
        </td>
        <td
          class="sc-gKsewC gbAxPW"
          data-testid="votes-Revenge of the Sith"
          headers="total-votes"
        >
          0
        </td>
      </tr>
    </tbody>
  </table>
  <p
    class="sc-jrAGrp gLqfYD"
  >
    Total Votes: 
    <span
      class="sc-kEjbxe"
    >
      0
    </span>
  </p>
</div>
`;

exports[`<App/> component shows three possible components shows a loading screen before returning API data 1`] = `
<p
  class="sc-jrAGrp gLqfYD"
  data-testid="loading"
>
  I find your lack of faith disturbing.  The app is 
  <strong>
    Loading...
  </strong>
</p>
`;

exports[`<App/> component shows three possible components shows an error screen when erroring 1`] = `
<p
  class="sc-jrAGrp gLqfYD"
  data-testid="api-error"
>
  You would be wise to surrender. An error has occurred: 
</p>
`;
